### PR TITLE
DOCS-463: Hierarhical dictionaries

### PR DESCRIPTION
**Pull request by BayoNet**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Documentation (changelog entry is not required)

### CODE REVIEW

File: docs/en/query_language/dicts/external_dicts_dict_hierarchical.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+----|--------------|------
+1 | 0 | Russia
+2 | 1 | Moscow
+3 | 2 | Center
+4 | 0 | Great Britain
+5 | 4 | London
+
+This table contains an attribute that equals the key of the nearest ancestor for the element.
+
+You can configure an hierarchical dictionary by setting the [hierarchical] (external_dict_dict_structure.md#hierarchical-dict-attr) property for the attribute that must contain the ancestor key.
```
Comment (qoega) [inline docs/en/query_language/dicts/external_dicts_dict_hierarchical.md]: I can't understand the logic. And it is too complicated.
configure..setting..property..attribute..must contain..ancestor.

Probably better tell it in several sentences. To enable ...  you should do following: 
- Table should have ancestor column with name `zzzz`.
- Parameter xxx is set to true.

Actually I can't tell how I should name columns in table for hierarchical dict to work. I don't see exact name here and there is no such parameter.

File: docs/en/query_language/dicts/external_dicts_dict_hierarchical.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+│       └── 3 (Center)  
+│
+└── 4 (Great Britain)
+    │
+    └── 5 (London)
+```
+
+This hierarchy can be expressed as the following dictionary table.
+
+Key | Ancestor key | Name
```
Comment (qoega) [inline docs/en/query_language/dicts/external_dicts_dict_hierarchical.md]: I believe it is easier for everyone to use parent-child terminology.

File: docs/ru/query_language/functions/ext_dict_functions.md [hunk 3]
```diff
@@ -111,24 +111,27 @@ dictHas('dict_name', id)
 
 Тип — `UInt8`.
 
-## dictGetHierarchy
+## dictGetHierarchy {#dictgethierarchy}
 
-Для иерархического словаря возвращает массив ключей словаря, начиная с переданного `id_expr` и продолжая цепочкой родительских элементов.
+Создаёт массив, содержащий цепочку предков для заданного ключа в [иерархическом словаре](../dicts/external_dicts_dict_hierarchical.md).
```
Comment (qoega) [inline docs/ru/query_language/functions/ext_dict_functions.md]: Возвращает?

File: docs/ru/query_language/dicts/external_dicts_dict_structure.md [hunk 4]
```diff
@@ -109,8 +127,12 @@ ClickHouse поддерживает следующие виды ключей:
...
-| `hierarchical` | Поддержка иерархии. Отображение в идентификатор родителя.<br/><br/>Значение по умолчанию: `false`. | Нет |
-| `injective` | Признак [инъективности](https://ru.wikipedia.org/wiki/Инъекция_(математика)) отображения `id -> attribute`. <br/>Если `true`, то обращения к словарям с включенной инъективностью могут быть автоматически переставлены ClickHouse за стадию `GROUP BY`, что как правило существенно сокращает их количество.<br/><br/>Значение по умолчанию: `false`. | Нет |
+<a name="hierarchical-dict-attr"></a> `hierarchical` | Если `true`, то атрибут содержит ключ предка для текущего элемента. Смотрите [Иерархические словари](external_dicts_dict_hierarchical.md).<br/><br/>Default value: `false`. | No| `injective` | Признак [инъективности](https://ru.wikipedia.org/wiki/Инъекция_(математика)) отображения `id -> attribute`. <br/>Если `true`, то обращения к словарям с включенной инъективностью могут быть автоматически переставлены ClickHouse за стадию `GROUP BY`, что как правило существенно сокращает их количество.<br/><br/>Значение по умолчанию: `false`. | Нет |
```
Comment (qoega) [inline docs/ru/query_language/dicts/external_dicts_dict_structure.md]: why do we need here html tags?
<a name="hierarchical-dict-attr"></a>

File: docs/ru/query_language/functions/ext_dict_functions.md [hunk 5]
```diff
@@ -111,24 +111,27 @@ dictHas('dict_name', id)
...
 ```sql
-dictGetHierarchy('dict_name', id)
+dictGetHierarchy('dict_name', key)
 ```
 
 **Параметры**
 
 - `dict_name` — имя словаря. [Строковый литерал](../syntax.md#syntax-string-literal).
-- `id_expr` — значение ключа словаря. [Выражение](../syntax.md#syntax-expressions), возвращающее значение типа [UInt64](../../data_types/int_uint.md).
+- `key` — значение ключа. [UInt64](../../data_types/int_uint.md).
```
Comment (qoega) [inline docs/ru/query_language/functions/ext_dict_functions.md]: If I understand correctly expression is not just a single value.

**Review by qoega — COMMENTED** [review]
After reading this I can't answer the question how to create hierarchical dict by myself and how to use them.

File: docs/ru/query_language/dicts/external_dicts_dict_structure.md (hunk 4 above)
Comment (BayoNet) [reply docs/ru/query_language/dicts/external_dicts_dict_structure.md]: <br> - New line inside the table cell.
<a name="hierarchical-dict-attr"></a> - ankor. It allows to refer directly to the table string. Can be placed in any position in the text.

File: docs/en/query_language/dicts/external_dicts_dict_hierarchical.md (hunk 2 above)
Comment (BayoNet) [reply docs/en/query_language/dicts/external_dicts_dict_hierarchical.md]: Ok.

File: docs/en/query_language/dicts/external_dicts_dict_hierarchical.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+----|--------------|------
+1 | 0 | Russia
+2 | 1 | Moscow
+3 | 2 | Center
+4 | 0 | Great Britain
+5 | 4 | London
+
+This table contains an attribute that equals the key of the nearest ancestor for the element.
```
[the remark this answers is not in the document]
Comment (BayoNet) [reply docs/en/query_language/dicts/external_dicts_dict_hierarchical.md]: This is terminology for dictionaries. But in this case the column looks as more appropriate term.
Ok.

File: docs/en/query_language/dicts/external_dicts_dict_hierarchical.md (hunk 1 above)
Comment (BayoNet) [reply docs/en/query_language/dicts/external_dicts_dict_hierarchical.md]: I've rewritten all this part.

File: docs/ru/query_language/functions/ext_dict_functions.md (hunk 3 above)
Comment (BayoNet) [reply docs/ru/query_language/functions/ext_dict_functions.md]: Создаёт и возвращает. Первичное действие всё же "создаёт".

File: docs/ru/query_language/dicts/external_dicts_dict_structure.md (hunk 4 above)
Comment (qoega) [inline docs/ru/query_language/dicts/external_dicts_dict_structure.md]: предок->родитель?

Comment (BayoNet) [reply docs/ru/query_language/dicts/external_dicts_dict_structure.md]: Вроде как предок-потомок.
https://ru.wikipedia.org/wiki/Иерархическая_модель_данных

**Pull request merged**